core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -578,7 +578,7 @@ class Partition(val topicPartition: TopicPartition,
 
   // Returns a verification guard object if we need to verify. This starts or continues the verification process. Otherwise return null.
   def maybeStartTransactionVerification(producerId: Long): Object = {
-    leaderLogIfLocal match {
+    log match {
```

Comment:
Should we add a test to avoid regressing in the future?

Reply:
This was found due to a specific race condition where leadership was elected in the middle of handling the produce request. I can try to recreate.

Reply:
What are the cases when log is none?  Could we get into a situation when the log is none at the time of this call and we decide to skip verification, but by the time we actually do produce the log is set up?

I think it would be more robust to require leadership at this point and return an error if it's not a leader yet.  We don't expect to successfully produce to a non-leader, so the error seams reasonable.

Reply:
I am a bit concerned with changing the flow here though. At the very least I suspect that some produce requests will be slower since we will have to retry in some cases when we elected. Maybe this isn't a big concern, but I still think it is a larger change.

As for the log not existing -- would we elect a log as leader and send a request to it before the log exists? 

@hachikuji what do you think?

Reply:
Returning an error seems straightforward. No harm done if the broker comes leader just after since the client will retry. Given how hard the scenario was to reproduce, I'm not too concerned about impact from retries.